### PR TITLE
#109/digest

### LINKED ISSUE #109: `DigestPage.tsx` 🍐
### Expected Behavior

Users need to be able to navigate to a new `Digest` page when they are logged in.

The page should display a thumbnail for each event that corresponds to one of the last 7 days in the current project.

Image

### Current Behavior

n/a - this is a new feature

### Any other notes

The most straightforward way of achieving this feature is probably to evaluate the `current_index` and populate the page with a card for each of the days that have passed in the project.

### Assignment

This issue is free for anyone to take

### PR DESCRIPTION
**Pull request by camelPhonso**
# Description

**Closes #109**  

This PR implements the final feature - ✨ Digest Mode ✨

### Files changed

 - `page-components/DigestPage/**`, `store.ts`, `popup.tsx` && `BurgerMenu.tsx` - new page added and navigation extended to give the user access to it from the burger menu
 - `components/EventCard/**` - new component to display the title and a thumbnail for each event in the digest page
 - `utils/determineActiveProjectId.ts` - new util to package logic evaluating the right project Id to target in a query
 - `eventAlarmListener.ts` && `index.ts` - implement new storage value `arebyte-digest-count` to track the user's progress in the digest feature
 - `eventTypes.ts` - new interface for the returned value for digest feature
 - `background/messages/fetchProjectDigest.ts` - new message handler returning an array of events for the digest page
 - `utils/generatePagination.ts` && `queries/eventDigestQuery.ts` - new utils processing the current digest count and creating a query to returned the filtered events for digest

### UI changes

Screenshot 2024-11-07 at 17 09 43

Screenshot 2024-11-07 at 17 09 26

### Changes to Documentation

n/a

# Tests

- `generatePagination.test.js` - unit tests for new util

### CODE REVIEW

File: src/types/eventTypes.ts [hunk 1]
```diff
@@ -71,3 +71,9 @@ export interface SlimPopup {
   description: BlocksContent
   text_content?: BlocksContent
 }
+
+export interface DigestedEvent {
+  id: number
+  title: string
+  pop_ups: Favourite[]
+}
```
Comment (nichgalzin) [inline src/types/eventTypes.ts:80]: I don't think we need a new type for this. This is just EventData interface minus a couple of fields.

File: src/types/projectTypes.ts [hunk 2]
```diff
@@ -31,3 +31,8 @@ export interface TagData {
   id: number
   name: string
 }
+
+export interface FullProject extends Omit<ProjectData, "sequence"> {
```
Comment (nichgalzin) [inline src/types/projectTypes.ts:35]: I think I'm a little confused about this interface there too. Let's talk about but seems maybe like we're doing a lot of work here to omit a couple of fields from the sequence interface

File: src/background/messages/fetchProjectDigest.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import { UserSession } from "~types/userTypes"
+import determineActiveProjectId from "~utils/determineActiveProjectId"
+import { fetchStrapiContent } from "~utils/fetchStrapiContent"
+import generatePagination from "~utils/generatePagination"
+import newStorage from "~utils/newStorage"
+
+const handler: PlasmoMessaging.MessageHandler = async (req, res) => {
+  const storage = newStorage()
+  const userSession: UserSession = await storage.get(
+    "arebyte-audience-session"
```
Comment (nichgalzin) [inline src/background/messages/fetchProjectDigest.ts]: Now with all the fighting we've been doing with storage I wonder if we want to store this in the user content type and then just fetch it in our initial user fetch and store it in the user store context object.

File: src/background/messages/fetchProjectDigest.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  )
+  const digestDayCount: number = await storage.get(
+    "arebyte-digest-count"
+  )
+
+  const activeProjectId = await determineActiveProjectId(
```
Comment (nichgalzin) [inline src/background/messages/fetchProjectDigest.ts]: I think I'm a little confused about the need for this db call. Won't this just be the current project I'd in the user context store?

File: src/utils/determineActiveProjectId.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import { CurrentProjectData } from "~types/projectTypes"
+
+import { fetchStrapiContent } from "./fetchStrapiContent"
+
+/**
+ *
+ * @summary evaluates the project_id saved in storage and returns the correct Id of the project that should be targeted in a fetch call
+ * @param givenId expects the project_id returned in the stored userSession
```
Comment (nichgalzin) [inline src/utils/determineActiveProjectId.ts]: See comment about in message handler

File: src/utils/eventAlarmListener.ts [hunk 6]
```diff
@@ -30,6 +33,10 @@ export default async function eventAlarmListener(alarm) {
       ...(newIndex === 0 && { project_id: 0 })
     })
     await storage.set("arebyte-audience-session", updatedSession)
+    await storage.set(
```
Comment (nichgalzin) [inline src/utils/eventAlarmListener.ts]: Think we might need to refactor this in light of what we were working on this morning. Think this should all move the db

**Review by nichgalzin — CHANGES_REQUESTED** [review]

File: src/background/messages/fetchProjectDigest.ts (hunk 4 above)
Comment (camelPhonso) [reply src/background/messages/fetchProjectDigest.ts]: yep - now getting this data directly from the cms

File: src/types/projectTypes.ts (hunk 2 above)
Comment (camelPhonso) [reply src/types/projectTypes.ts:35]: keeping this because you were a silly sausage

File: src/background/messages/fetchProjectDigest.ts (hunk 3 above)
Comment (camelPhonso) [reply src/background/messages/fetchProjectDigest.ts]: cool I've now refactored the feature to reference a new field in the cms after merging the PR with the new data flow 👍

File: src/utils/eventAlarmListener.ts (hunk 6 above)
Comment (camelPhonso) [reply src/utils/eventAlarmListener.ts]: done :)

**Review by nichgalzin — APPROVED** [review]
Bang-o-Rang!

**Pull request merged**